mapper/orm/src/main/java/org/hibernate/search/mapper/orm/outbox/impl/OutboxEventBackgroundExecutor.java

```diff
@@ -0,0 +1,283 @@
+ */
+package org.hibernate.search.mapper.orm.outbox.impl;
+
+import static org.hibernate.search.mapper.orm.common.impl.TransactionUtils.withinTransaction;
+
+import java.lang.invoke.MethodHandles;
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Future;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicReference;
+
+import org.hibernate.Session;
+import org.hibernate.query.Query;
+import org.hibernate.search.engine.backend.orchestration.spi.SingletonTask;
+import org.hibernate.search.engine.reporting.EntityIndexingFailureContext;
+import org.hibernate.search.engine.reporting.FailureHandler;
+import org.hibernate.search.mapper.orm.automaticindexing.spi.AutomaticIndexingMappingContext;
+import org.hibernate.search.mapper.orm.logging.impl.Log;
+import org.hibernate.search.mapper.pojo.route.DocumentRouteDescriptor;
+import org.hibernate.search.mapper.pojo.route.DocumentRoutesDescriptor;
+import org.hibernate.search.util.common.logging.impl.LoggerFactory;
+import org.hibernate.search.util.common.serialization.spi.SerializationUtils;
+
+public class OutboxEventBackgroundExecutor {
+
+	private static final Log log = LoggerFactory.make( Log.class, MethodHandles.lookup() );
+	private static final int MAX_RETRIES = 3;
+
+	private enum Status {
+		STOPPED,
+		STARTED
+	}
+
+	private final AutomaticIndexingMappingContext mapping;
+	private final int pollingInterval;
+	private final int batchSize;
+	private final AtomicReference<Status> status = new AtomicReference<>( Status.STOPPED );
+	private final FailureHandler failureHandler;
+	private final SingletonTask processingTask;
+
+	public OutboxEventBackgroundExecutor(AutomaticIndexingMappingContext mapping, ScheduledExecutorService executor,
+			int pollingInterval, int batchSize) {
+		this.mapping = mapping;
+		this.pollingInterval = pollingInterval;
+		this.batchSize = batchSize;
+
+		failureHandler = mapping.failureHandler();
+		processingTask = new SingletonTask(
+				"Delayed commit for " + OutboxTableAutomaticIndexingStrategy.NAME,
+				new HibernateOrmOutboxWorker(),
+				new HibernateOrmOutboxScheduler( executor ),
+				failureHandler
+		);
+	}
+
+	public void start() {
+		status.set( Status.STARTED );
+		processingTask.ensureScheduled();
+	}
+
+	public CompletableFuture<?> completion() {
+		status.set( Status.STOPPED );
+		return processingTask.completion();
+	}
+
+	public void stop() {
+		processingTask.stop();
+	}
+
+	private class HibernateOrmOutboxWorker implements SingletonTask.Worker {
+
+		@Override
+		public CompletableFuture<?> work() {
+			if ( mapping.sessionFactory().isClosed() ) {
+				log.sessionFactoryIsClosedOnOutboxProcessing();
+				return CompletableFuture.completedFuture( null );
+			}
```

Comment:
Tests still pass if I remove this block of code. Please do?

As I explained in another comment, I don't expect this method to be called when `mapping.sessionFactory().isClosed()` returns true.

Reply:
Running `OutboxTableAutomaticIndexingStrategyBaseIT` we got several of these calls.
<img width="1858" alt="Screenshot 2021-03-29 at 14 19 33" src="https://user-images.githubusercontent.com/1748775/112835661-dbfb2f00-9099-11eb-803b-634af6112c94.png">

Reply:
Indeed. I see there are stacktraces if I remove this code, though the tests don't seem to fail.

From what I can see, the session factory marks itself as closed before it calls `sessionFactoryClosing` on its observers, which leads to all queries throwing exceptions.

I opened https://hibernate.atlassian.net/browse/HHH-14541. Please add a comment referencing it? We really need to fix that bug, because this check won't save us if the session factory gets closed when we're right in the middle of an execution.